homeworks/01-bubble-sort/src/components/app/App.tsx

```diff
@@ -0,0 +1,166 @@
+  paused = "paused",
+  solved = "solved"
+}
+
+interface IState {
+  elements: number[],
+  delay: number, // ms
+  status: Status,
+  elemsCountError: boolean,
+  elemsDelayError: boolean
+}
+
+class App extends React.Component<{}, IState> {
+  state: IState = {
+    elements: Array(20).fill(0),
+    delay: 50,
+    status: Status.notStarted,
+    elemsCountError: false,
+    elemsDelayError: false
+  };
+
+  private setElements(elements: number[]): void {
+    this.setState((state: IState) => ({ ...state, elements }));
+  }
+
+  private setRandomElements(count: number = this.state.elements.length): void {
+    this.setState((state: IState) => ({
+      ...state,
+      elemsCountError: false,
+      elemsDelayError: false,
+      status: Status.notStarted,
+      elements: randomArray(count)
+    }));
+  }
+
+  private newSetHandler = () => {
+    this.setState({ ...this.state, status: Status.notStarted });
+    this.setRandomElements();
+  }
+
+  private setStatus(s: Status) {
+    this.setState({ ...this.state, status: s });
+  }
+
+  private sortHandler = async (event: React.MouseEvent<HTMLButtonElement>) => {
+    switch (this.state.status) {
+      case Status.started:
+        this.setStatus(Status.paused)
+        return;
+      case Status.paused:
+        this.setStatus(Status.started)
+        return;
+      case Status.solved:
+        this.setStatus(Status.solved)
+        return;
+      default:
+        this.setStatus(Status.started)
+    }
+
+    for (let lastIndex: number = this.state.elements.length - 1; lastIndex >= 1 ; lastIndex-- ) {
+      let changed: boolean = false;
+
+      for (let i: number = 0; i < lastIndex; i++) {
+        const els: number[] = [...this.state.elements];
+        if (els[i] > els[i + 1]) {
+          [els[i + 1], els[i]] = [els[i], els[i + 1]];
+          this.setElements(els);
+          changed = true;
+        }
+
+        await delay(this.state.delay);
+
+        if (this.state.status === Status.notStarted) {
+          this.setRandomElements();
+          return;
+        }
+
+        if (this.state.status === Status.paused) i--;
+      }
+
+      if (!changed) {
+        this.setStatus(Status.solved)
+        return;
+      }
+    }
+
+    this.setStatus(Status.solved)
+  }
+
+  private changeCountHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const val: number = Number.parseInt(event.target.value);
+      if (!isNaN(val) && val > 1 && val <= 100) {
+        this.setRandomElements(val);
+      } else this.setState({ ...this.state,  elemsCountError: true });
+  }
+
+  private changeDelayHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const val: number = Number.parseInt(event.target.value);
+      if (!isNaN(val) && val > 1 && val <= 5000) {
+        this.setState({ ...this.state,  delay: val, elemsDelayError: false })
+      } else this.setState({ ...this.state,  elemsDelayError: true });
+  }
+
+  componentDidMount(): void {
+    this.setRandomElements();
```

Comment:
Guess, it will be one more redundant re-render. You can set it once during component initialization.